ARROW-12017: [R] [Documentation] Make proper developing arrow docs

jonkeane opened the pull request.

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

nealrichardson commented on r/vignettes/dev-docs.Rmd:
I believe this is only relevant for macOS, so you could make this conditional (since you're already doing conditional things)

jonkeane replied:
Yeah, I will do that + the other various conditionals when I start writing the actual docs. I wanted to get it building for now and then edit/add to it knowing that everything I was writing was tested later.

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,349 @@
+zip file before installing the `arrow` R package. Version numbers in that
+repository correspond to dates, and you will likely want the most recent.
+
+## Developer envorinment setup
```
```suggestion
## Developer environment setup
```

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,349 @@
+First, install the C++ library. See the [developer
+guide](https://arrow.apache.org/docs/developers/cpp/building.html) for details. 
+
+### Install dependencies {.tabset}
```
what is `{.tabset}`?

jonkeane replied:
We might not want to take it on, but with https://github.com/apache/arrow/blob/84771514e05db8ef959f6cdfacb4aeb337a55bad/r/pkgdown/extra.js it will make the next heading down tabbed when the pkgdown site renders. 

To see it in action (without pulling this branch + compling pkgdown yourself), it'll be similar to how the different SQL dialects have tabs two code blocks down at https://dittodb.jonkeane.com/#a-quick-example

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,349 @@
+
+### Install dependencies {.tabset}
+
+These dependencies are technically only needed for S3 support(?)
```
Yes, only S3 support.

Otherwise Arrow doesn't require any other system dependencies (beyond what R itself requires)--however, if you're providing a build script that you want a dev to run separately, probably should have them install `cmake`.

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,349 @@
+
+### Building Arrow {.tabset}
+
+It’s recommended to make a build directory inside of the cpp directory of the Arrow git repository (it is git-ignored). Assuming you are inside cpp/build, you’ll first call cmake to configure the build and then make install. For the R package, you’ll need to enable several features in the C++ library using -D flags:
```
```suggestion
It’s recommended to make a `build` directory inside of the `cpp` directory of the Arrow git repository (it is git-ignored). Assuming you are inside `cpp/build`, you’ll first call `cmake` to configure the build and then `make install`. For the R package, you’ll need to enable several features in the C++ library using `-D` flags:
```

`make install` is correct because the default generator for cmake is `Unix Makefiles`. You aren't (presumably) going to get into using alternative generators here (ninja etc.), but if you wanted to have the command that will work regardless of whether you use ninja, msvc, etc., it would be `cmake --build . --target install`

nealrichardson commented on r/vignettes/dev-docs.Rmd:
````diff
@@ -0,0 +1,349 @@
+export LIB_DIR=$ARROW_HOME/lib
+```
+
+TODO: how to add these vars separated by colons if they have other things in them already (eg LD_LIBRARY_PATH=$(pwd)/dist/lib:$LD_LIBRARY_PATH)
````
Yes, you definitely don't want to overwrite LD_LIBRARY_PATH like your code currently does, I think that can cause unexpected breakage.

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,349 @@
+
+TODO: how to add these vars separated by colons if they have other things in them already (eg LD_LIBRARY_PATH=$(pwd)/dist/lib:$LD_LIBRARY_PATH)
+
+On linux, you will need to set `LD_LIBRARY_PATH` to this same directory before launching R and using Arrow. One way to do this is to add it to your profile. # TODO: do we want to recommend this? Is there any other alternative?
```
I think you also want to explain why you wouldn't need to do this on macOS (something something rpath).

Alternatives: probably setting something in R's Makevars

jonkeane replied:
I don't think that Makevars will do it, since this is important at runtime not at build time, but I'll try it and see.

nealrichardson commented on r/vignettes/dev-docs.Rmd:
````diff
@@ -0,0 +1,349 @@
+If you would like to use arrow from an alternative directory
+
+```{bash, save=run & !sys_install}
+export ARROW_HOME=$(pwd)/dist
````
I think you need to explain where "dist" comes from (and note that it is arbitrary, you can call it whatever you want)

nealrichardson commented on r/vignettes/dev-docs.Rmd:
````diff
@@ -0,0 +1,349 @@
+
+```{bash, save=run & !sys_install}
+export ARROW_HOME=$(pwd)/dist
+export LD_LIBRARY_PATH=$(pwd)/dist/lib
````
```suggestion
export LD_LIBRARY_PATH="${ARROW_HOME}/lib:$LD_LIBRARY_PATH"
```

nealrichardson commented on r/vignettes/dev-docs.Rmd:
````diff
@@ -0,0 +1,349 @@
+```{bash, save=run & !sys_install}
+export ARROW_HOME=$(pwd)/dist
+export LD_LIBRARY_PATH=$(pwd)/dist/lib
+export INCLUDE_DIR=$ARROW_HOME/include
+export LIB_DIR=$ARROW_HOME/lib
````
We could add handling to `r/configure` that looks for `$ARROW_HOME` and sets these accordingly. One downside to setting INCLUDE_DIR and LIB_DIR to the arrow install dir is that some other packages' configure scripts use the same env vars, and (e.g.) curl's headers/libs are not in those directories. So it would be better to have an arrow-specific var IME.

jonkeane replied:
Yeah, that makes sense — part of what I was hoping this would do is show places like that. I think that makes sense + will add it so that we don't have to depend on those specifically.

nealrichardson commented on r/vignettes/dev-docs.Rmd:
````diff
@@ -0,0 +1,349 @@
+  ..
+```
+
+#### Installing to the system
````
I would lead with this one and then what you're doing is adding `-DCMAKE_INSTALL_PREFIX=$ARROW_HOME`

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,349 @@
+  -DARROW_WITH_SNAPPY=ON \
+  -DARROW_WITH_ZLIB=ON \
+  -DARROW_INSTALL_NAME_RPATH=OFF \
+  -DThrift_SOURCE=BUNDLED \
```
I would leave this off and suggest it later in a troubleshooting section

jonkeane replied:
Yeah, I've since removed this

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,349 @@
+  -DARROW_JEMALLOC=ON \
+  -DARROW_JSON=ON \
+  -DARROW_PARQUET=ON \
+  -DCMAKE_BUILD_TYPE=release \
```
I'd leave this one off of the main list too; release is default. You can suggest "debug" or "relwithdebinfo" as options for later (you don't want them normally because they're much slower, but sometimes you need them if you're debugging)

nealrichardson reviewed (commented):
A few notes. Can you also delete the corresponding content from README.md?

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
@github-actions crossbow submit test-r-devdocs

jonkeane commented:
OK, I think that I've addressed all of the comments except for https://github.com/apache/arrow/pull/9898/files#r610775175 (though the GH interface + the number of commits here doesn't make it easy to ensure that I have!) 

> Can you also delete the corresponding content from README.md?
We will, though the plan is to do this after both this PR and ARROW-11477 are completed so we don't conflict + don't accidentally remove something without it going in one of the new locations first.

> I would lead with [system install] and then what you're doing is adding -DCMAKE_INSTALL_PREFIX=$ARROW_HOME

Since this section and the one above it are under {.tabset} they will be tabbed, such that one or the other will be displayed on pkgdown (in the local-vignette* it will be in this order, but I suspect that the vast majority of people looking at these look at the pkgdown site). Since this is the case, for either we will have the full `cmake` command such that each tab is a full install, just of the two different types.

I'm a bit torn as to if we should recommend system-install or user-dir-install. I've listed the pros/cons for system-install but they are the mirror image, so flip them for the pros/cons for user-dir-install.

system-install pros: 
  * easier in many cases (don't need to deal with `LD_LIBRARY_PATH` on linux)
  * more natural install for R

system-install cons:
  * clobbers system installs of Arrow
  * can only have one libarrow around at a time
  * isn't aligned with python or c++ docs
  
I'm leaning towards recommending the user-dir-install to align with other developer docs. We could also suggest building bundled builds where we bundle the needed Arrow library dependencies into the R package. I'm hesitant to go in that direction because either one will have to rebuild libarrow each time the r package is installed, or maintain (and prevent from being deleted) the pre-built libarrow under `arrow/r/libarrow`. We might be able to put that in a different place and recommend that, but that starts being very similar to the user-dir-install approach.
 
* - The local-vignette *does* strip out the {.tabset} attribute, but it doesn't support tabs so they aren't there.

ianmcook commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+
+The `arrow` package uses some customized tools on top of `cpp11` to
+prepare its C++ code in `src/`. If you change C++ code in the R package,
+you will need to set the `ARROW_R_DEV` environment variable to `TRUE`
```
```suggestion
you will need to set the `ARROW_R_DEV` environment variable to `true`
```

ianmcook commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+* All tests are skipped on Linux if the package builds without the C++ libarrow.
+  To make the build fail if libarrow is not available (as in, to test that
+  the C++ build was successful), set `TEST_R_WITH_ARROW=TRUE`
+* Some tests are disabled unless `ARROW_R_DEV=TRUE`
```
for consistency
```suggestion
* Some tests are disabled unless `ARROW_R_DEV=true`
```

nealrichardson commented on r/configure:
```diff
@@ -66,8 +66,12 @@ if [ "$FORCE_AUTOBREW" = "true" ] || [ "$FORCE_BUNDLED_BUILD" = "true" ]; then
+  echo "*** Using ARROW_HOME as the source of libarrow"
+  PKG_CFLAGS="-I$ARROW_HOME/include $PKG_CFLAGS"
+  PKG_DIRS="-L$ARROW_HOME/lib"
+elif[ "$INCLUDE_DIR" ] || [ "$LIB_DIR" ]
```
```suggestion
elif [ "$INCLUDE_DIR" ] || [ "$LIB_DIR" ]
```
also I've never understood why this was valid as an `||` instead of `&&` 🤷

jonkeane replied:
Hmm, yeah that should be && lemme see what I can do...

`elif [ "$INCLUDE_DIR" ] && [ "$LIB_DIR" ]; then` seems to work

ianmcook commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+
+* All tests are skipped on Linux if the package builds without the C++ libarrow.
+  To make the build fail if libarrow is not available (as in, to test that
+  the C++ build was successful), set `TEST_R_WITH_ARROW=TRUE`
```
```suggestion
  the C++ build was successful), set `TEST_R_WITH_ARROW=true`
```

ianmcook commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+  the C++ build was successful), set `TEST_R_WITH_ARROW=TRUE`
+* Some tests are disabled unless `ARROW_R_DEV=TRUE`
+* Tests that require allocating >2GB of memory to test Large types are disabled
+  unless `ARROW_LARGE_MEMORY_TESTS=TRUE`
```
```suggestion
  unless `ARROW_LARGE_MEMORY_TESTS=true`
```

ianmcook commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+this by installing LLVM via Homebrew and running the script as
+`CLANG_FORMAT=$(brew --prefix llvm@8)/bin/clang-format ./lint.sh`
+
+_note_ that the lint script requires python3 and the python dependencies (note that `cmake_format is pinned to a specific version):
```
```suggestion
_Note_ that the lint script requires Python 3 and the Python dependencies (note that `cmake_format is pinned to a specific version):
```

ianmcook replied:
Also, do you think it's worth adding the bit about possibly needing to change the Python hashbangs on the first line of `cpp/build-support/run_clang_format.py` and `cpp/build-support/run_cpplint.py` to point to Python 3?

jonkeane replied:
I think the right ™️ way to do that is with virtualenvs created with python3. We can add that, but I think at that point we should write that at a higher level in the arrow repo and link to it.

ianmcook commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+
+We use Google C++ style in our C++ code. Check for style errors with
+
+    ./lint.sh
```
`````suggestion
``` shell
./lint.sh
```
`````

ianmcook commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+
+Fix any style issues before committing with
+
+    ./lint.sh --fix
```
`````suggestion
``` shell
./lint.sh --fix
```
`````

nealrichardson commented on r/pkgdown/extra.js:
```diff
@@ -0,0 +1,166 @@
+/**
+ * adapted from rmarkdown/inst/rmd/h/navigation-1.1/tabsets.js
```
Why do we have to copy this from rmarkdown? Does `rmarkdown` not make it available? Can we load https://raw.githubusercontent.com/rstudio/rmarkdown/master/inst/rmd/h/navigation-1.1/tabsets.js at page load time in the pkgdown site?

I'm not sure we can do this. [`rmarkdown` is GPL-3 license](https://github.com/rstudio/rmarkdown/blob/master/DESCRIPTION#L113), which is copyleft and on the [forbidden list](https://www.apache.org/legal/resolved.html#category-x). We can of course *use* `rmarkdown`, we just can't copy its code inside of apache/arrow.

jonkeane replied:
pkgdown doesn't pull this code in to the javascript it loads, so it doesn't contain it itself (I've thought about sending a PR to add it there, [though they seemed down on the idea generally](https://github.com/r-lib/pkgdown/issues/981))

This code isn't _exactly_ what rmarkdown uses since the pkgdown structure is slightly different than the standard rmarkdown html template.

The [underlying jquery plugin](https://github.com/aidanlister/jquery-stickytabs) is MIT licensed, the `buildTabsets` is the only code that is coming from rmarkdown (the `two $(document.read(...)` functions I wrote as part of dittodb).

jonkeane replied:
Ok, I think I've resolved this, we now get that javascript file from GH (though we have to use this JS cdn to get it with the correct mimetype).

Then I managed to work some magic that edits the function like I needed in place. As far as I can see building it locally it works exactly like I expect it should.

nealrichardson commented on r/vignettes/dev-docs.Rmd:
```diff
@@ -0,0 +1,427 @@
+---
```
I might change the file name to something like `developing.Rmd`, makes for a better URL IMO.

nealrichardson commented on r/vignettes/developing.Rmd:
````diff
@@ -0,0 +1,457 @@
+brew install apache-arrow --HEAD
+```
+
+On Windows, you can download a .zip file with the arrow dependencies from the
````
This is also true for Linux now

nealrichardson commented on r/vignettes/developing.Rmd:
````diff
@@ -0,0 +1,457 @@
+```
+
+On Windows, you can download a .zip file with the arrow dependencies from the
+[nightly repository](https://arrow-r-nightly.s3.amazonaws.com/libarrow/bin/windows/),
````
~~Unfortunately~~ As it turns out, this URL does not return useful information (the bintray version did). Surely there's an S3 permission that can be set that would make that list (XML) with contents, but a better way is probably to encourage use of the package itself, like: 

```
> nightly <- s3_bucket("arrow-r-nightly")
> nightly$ls("libarrow/bin")
[1] "libarrow/bin/centos-7"     "libarrow/bin/centos-8"    
[3] "libarrow/bin/debian-10"    "libarrow/bin/debian-9"    
[5] "libarrow/bin/ubuntu-16.04" "libarrow/bin/ubuntu-18.04"
[7] "libarrow/bin/windows"
```

We could even add a function to download the latest binary (appropriate for your OS) and unzip it/put it in the right place. Out of scope here, just saying.

jonkeane replied:
> We could even add a function to download the latest binary (appropriate for your OS) and unzip it/put it in the right place. Out of scope here, just saying.

I'll add this to the todo Jira, it's a good idea and something I had struggled with a little bit myself when trying to diagnose how the nightlies are downloaded and if we can match those / do that process manually/with a development version.

nealrichardson commented on r/vignettes/developing.Rmd:
````diff
@@ -0,0 +1,457 @@
+```
+
+
+_Special instructions on Linux:_ You will need to set `LD_LIBRARY_PATH` to the same directory as `LIB_DIR` before launching R and using Arrow. One way to do this is to add it to your profile (we use `~/.bash_profile` here, but you might need to put this in a different file depending on your setup). On macOS we do not need to do this because the macOS shared library paths are hardcoded to their lo ...
````
I don't think `LIB_DIR` is mentioned in this document so that's not helpful (anymore)

nealrichardson reviewed (commented):
I pushed some edits, and here are a few more comments. I haven't read through all the way to the end yet, will spend more time with this tomorrow. 

One thought is that we could make note at the beginning that this is a work in progress, we know it works for many cases (because we do this ourselves, and we test it in CI!), but as with any workflow thing, there are differences of opinion on what the best way to do things is, so we welcome suggestions/additions/pull requests.

thisisnic commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,465 @@
+
+### Build Arrow
+
+You can `-j#` here too to speed up compilation by running in parallel (where `#` is the number of cores you have available).
```
Tiny point, but took me a couple of reads to realise that this sentence means you can add `-j#` to the command below.

nealrichardson commented on dev/tasks/r/github.devdocs.yml:
```diff
@@ -0,0 +1,92 @@
+
+jobs:
+  devdocs:
+    name: "install from the devdoc instructions"
```
Can you make the name explicitly include the matrix values? As it currently reads, it just says true, false, and IDK what that means.

nealrichardson commented on r/configure:
```diff
@@ -66,8 +66,12 @@ if [ "$FORCE_AUTOBREW" = "true" ] || [ "$FORCE_BUNDLED_BUILD" = "true" ]; then
 # Note that cflags may be empty in case of success
-if [ "$INCLUDE_DIR" ] || [ "$LIB_DIR" ]; then
-  echo "*** Using INCLUDE_DIR/LIB_DIR"
+if [ "$ARROW_HOME" ]; then
```
This might merit a note in NEWS. In fact, the existence of the new vignette should be noted in NEWS.

nealrichardson commented on r/vignettes/install.Rmd:
```diff
@@ -244,30 +219,6 @@ Similarly, if you're using Arrow system libraries, running `update.packages()`
 after a new release of the `arrow` package will likely fail unless you first
 update the system packages.
 
-## Using a local Arrow C++ build
```
Good call, a user should not be here.

nealrichardson commented on r/vignettes/install.Rmd:
```diff
@@ -379,6 +330,13 @@ By default, these are all unset. All boolean variables are case-insensitive.
   The directory will be created if it does not exist.
 * `CMAKE`: When building the C++ library from source, you can specify a
   `/path/to/cmake` to use a different version than whatever is found on the `$PATH`
+* `ARROW_S3`: If set to `true` S3 support will be built (as long as the 
```
Let's make this a separate bullet list since these are features to be enabled/disabled in the cmake. Note also what the defaults are. Also might be good to enumerate the compression libraries (could group them all under `ARROW_WITH_*` or something).

jonkeane replied:
I also moved them up to the top of this section, I think these are actually slightly more interesting / important than the other group

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+Either way, you will need to create a directory into which the C++ build will put its contents. It is recommended to make a `build` directory inside of the `cpp` directory of the Arrow git repository (it is git-ignored, so you won't accidentally check it in).
+
+Starting from the directory that contains your git checkout of `apache/arrow`,
```
IMO we should make all instructions relative to the top-level of the repo, so drop the leading `arrow/` below

jonkeane replied:
I've tried to do that, generally with the caveat that the `dist` directory that is the `ARROW_HOME` should *not* be under this directory. 

IIRC, originally I had the creation of this directory right before the config (when the tutorial would be inside the repo) but you moved it up higher in your push. If we want the `mkdir call` to not include `arrow/` we should move it to a later code chunk after we've `pushd arrow`. 

I think it's probably better lower down anyway, so unless you object I think I'll move it back down there (along with most of the paragraph above).

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+### Configure the Arrow build {.tabset}
+
+You can choose to build and then install the Arrow library into a user-defined directory or into a system-level directory. You only need to do one of these two options.
```
Let's make clear that there are effectively three steps, which involve two directories:

* configure (`cmake`), which populates a build directory with scripts etc.
* build, which compiles arrow (including potentially downloading and building dependencies) and produces the resulting libraries and headers in that build directory
* install, which organizes just the libraries and headers into a clean location, without any of the build detritus.

It's this last step, where to install to, that is the key difference in the two approaches described here. And it requires a change or two in the configuring step.

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+#### Install to another directory
+
+It is recommended that you install the arrow library to a user-level directory to be used in development. In this example we will install it to a directory called `dist` that has the same parent as our `arrow` checkout, but it could be named or located anywhere you would like. However, note that your installation of the Arrow R package will point to this directory and need it to remain intact for ...
```
> It is recommended

Why is it recommended?

nealrichardson commented on r/vignettes/developing.Rmd:
````diff
@@ -0,0 +1,510 @@
+mkdir $ARROW_HOME
+```
+
+_Special instructions on Linux:_ You will need to set `LD_LIBRARY_PATH` to the `lib` directory that will is under where we set `$ARROW_HOME`, before launching R and using Arrow. One way to do this is to add it to your profile (we use `~/.bash_profile` here, but you might need to put this in a different file depending on your setup). On macOS we do not need to do this because the macOS shared libr ...
````
"if you use a shell other than `bash`"

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+Other flags that may be useful:
+
+* `-DARROW_EXTRA_ERROR_CONTEXT=ON` makes errors coming from the C++ library point to files and line numbers
```
Since this is a developer guide, should we always recommend this on?

jonkeane replied:
I've not actually used it, I can add it to the list if you think it's helpful

nealrichardson replied:
Yeah I think it's helpful

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+* `-DARROW_EXTRA_ERROR_CONTEXT=ON` makes errors coming from the C++ library point to files and line numbers
+* `-DBoost_SOURCE=BUNDLED` and `-DThrift_SOURCE=bundled`, for example, or any other dependency `*_SOURCE`, if you have a system version of a C++ dependency that doesn't work correctly with Arrow. This tells the build to compile its own version of the dependency from source.
+* `-DCMAKE_BUILD_TYPE=debug` and `-DCMAKE_BUILD_TYPE=relwithdebinfo` can be useful for debugging, though they are both slower to compile than the default `release`.
```
```suggestion
* `-DCMAKE_BUILD_TYPE=debug` or `-DCMAKE_BUILD_TYPE=relwithdebinfo` can be useful for debugging. You probably don't want to do this generally because a debug build is much slower at runtime than the default `release` build.
```

nealrichardson commented on r/vignettes/developing.Rmd:
````diff
@@ -0,0 +1,510 @@
+You can add `-j#` between `make` and `install` here too to speed up compilation by running in parallel (where `#` is the number of cores you have available).
+
+```{bash, save=run & !(sys_install & ubuntu)}
+make install
````
```suggestion
make -j8 install
```

Just to illustrate

jonkeane replied:
I think the most that we could do on CI is `-j2` (or `-j4` if there is multithreading, though I'm not certain there is son the runners)

nealrichardson replied:
Ok. I wasn't thinking at the time about running this on CI. That said, I just set MAKEVARS=-j100 locally and it didn't break 🤷

jonkeane replied:
Oh apparently saturation is less of a concern, so maybe we can do 8 and the CI will deal with swapping between them.

nealrichardson commented on r/vignettes/developing.Rmd:
````diff
@@ -0,0 +1,510 @@
+sudo make install
+```
+
+Note that after any change to the C++ library, you must reinstall it and
````
Maybe put this under Troubleshooting?

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+* Setting the environment variable `FORCE_BUNDLED_BUILD` to `true` will skip the `pkg-config` search for Arrow libraries and attempt to build from the same source at the repository+ref given.
+* You may also need to set the Makevars `CPPFLAGS` and `LDFLAGS` to `""` in order to prevent the installation process from attempting to link to already installed system versions of Arrow. One way to do this temporarily is wrapping your `remotes::install_github()` call like so: `withr::with_makevars(list(CPPFLAGS = "", LDFLAGS = ""), remotes::install_github(...))`.
+
+## How does our automated building actually work?
```
I don't think "automated building" is the right term here. Something about how all of the pieces come together? Or "what happens when you R CMD INSTALL?"

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+There are a number of scripts that are triggered when `R CMD INSTALL .`. For Arrow users, these should all just work without configuration and pull in the most complete pieces (e.g. official binaries that we host) so the installation process is easy. However knowing about these scripts can help troubleshoot if things go wrong in them or things go wrong in an install:
+
+* `configure` and `configure.win` These scripts are triggered during `R CMD INSTALL .` on non-windows and windows platforms respectively. They handle finding the Arrow library, setting up the build variables necessary, and writing the package Makevars file that is used to compile the C++ code in the R package.
```
Also note that this is a standard thing for R packages with compiled code, this is not an Arrow convention

```suggestion
* `configure` and `configure.win` These scripts are triggered during `R CMD INSTALL .` on non-Windows and Windows platforms, respectively. They handle finding the Arrow library, setting up the build variables necessary, and writing the package Makevars file that is used to compile the C++ code in the R package.
```

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+* `configure` and `configure.win` These scripts are triggered during `R CMD INSTALL .` on non-windows and windows platforms respectively. They handle finding the Arrow library, setting up the build variables necessary, and writing the package Makevars file that is used to compile the C++ code in the R package.
+* `tools/nixlibs.R` This script is called by `configure` on Linux (or on any non-windows OS with the environment variable `FORCE_BUNDLED_BUILD=true`). This sets up the build process for our bundled builds (which is the default on linux). The operative logic is at the end of the script, but it will do the following (and it will stop with the first one that succeeds and some of the steps are only c ...
+  * Check if there is an already built libarrow in `arrow/r/libarrow-{version}`, use that to link against if it exists.
+  * Check if a binary is available from our hosted official builds.
```
Not official

jonkeane replied:
That was a bad typo!

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+  * Check if a binary is available from our hosted official builds.
+  * Download the Arrow source and build the Arrow Library from source.
+  * `*** Proceed without C++` dependencies (this is an error and the package will not work, but if you see this message you know the previous steps have not succeeded/were not enabled)
+* `inst/build_arrow_static.sh` this script builds Arrow for a bundled, static build. It is called by `tools/nixlibs.R` when the Arrow library is being built.
```
Note that this should look familiar if you've read this far in this vignette--it's cmake and install.

(One difference, for another day, is static vs shared library)

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+There are a number of scripts that are triggered when `R CMD INSTALL .`. For Arrow users, these should all just work without configuration and pull in the most complete pieces (e.g. official binaries that we host) so the installation process is easy. However knowing about these scripts can help troubleshoot if things go wrong in them or things go wrong in an install:
+
+* `configure` and `configure.win` These scripts are triggered during `R CMD INSTALL .` on non-windows and windows platforms respectively. They handle finding the Arrow library, setting up the build variables necessary, and writing the package Makevars file that is used to compile the C++ code in the R package.
+* `tools/nixlibs.R` This script is called by `configure` on Linux (or on any non-windows OS with the environment variable `FORCE_BUNDLED_BUILD=true`). This sets up the build process for our bundled builds (which is the default on linux). The operative logic is at the end of the script, but it will do the following (and it will stop with the first one that succeeds and some of the steps are only c ...
```
It's not always called by configure

jonkeane replied:
👍 I've clarified this. Funny story: for me this phrasing doesn't imply "always" so my first reading of your comment was that some other thing calls `nixlibs.R` and I went looking for what else calls it until coming back and realizing what you meant. Oh langauge!

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+  * `*** Proceed without C++` dependencies (this is an error and the package will not work, but if you see this message you know the previous steps have not succeeded/were not enabled)
+* `inst/build_arrow_static.sh` this script builds Arrow for a bundled, static build. It is called by `tools/nixlibs.R` when the Arrow library is being built.
+
+## Editing C++ code
```
```suggestion
## Editing C++ code in the R package
```

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+## Editing C++ code
+
+The `arrow` package uses some customized tools on top of `cpp11` to
```
Why? (Something about how we have features that are conditionally built at compile time)

jonkeane replied:
TIL. Are there any other reasons I should add here too?

nealrichardson replied:
I think that's it.

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+The `arrow` package uses some customized tools on top of `cpp11` to
+prepare its C++ code in `src/`. If you change C++ code in the R package,
+you will need to set the `ARROW_R_DEV` environment variable to `true`
```
Or use the makefile

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+sessions) so that the `data-raw/codegen.R` file is used for code
+generation.
+
+We use Google C++ style in our C++ code. Check for style errors with
```
Honestly I'd suggest the editor plugins before asking people to run a shell script with python etc. etc.

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+* `@github-actions crossbow submit -g r` for extended R CI tests
+* `@github-actions crossbow submit {task-name}` for running a specific task
+* `@github-actions autotune` will run and fix lint c++ linting errors + run R documentation (among other cleanup tasks) and commit them to the branch
+
```
(at some point) we should also talk about docker and docker-compose

jonkeane replied:
I'm going to add this to a follow on (especially cause I think ultimately that probably should go on the project docs with a link here to that)

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+Crossbow is our extended test suite runner [more documentation is available](https://arrow.apache.org/docs/developers/crossbow.html) but important GitHub comment commands include:
+
+* `@github-actions crossbow submit -g r` for extended R CI tests
```
```suggestion
* `@github-actions crossbow submit -g r` for all extended R CI tests
```

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+Crossbow is our extended test suite runner [more documentation is available](https://arrow.apache.org/docs/developers/crossbow.html) but important GitHub comment commands include:
+
+* `@github-actions crossbow submit -g r` for extended R CI tests
+* `@github-actions crossbow submit {task-name}` for running a specific task
```
```suggestion
* `@github-actions crossbow submit {task-name}` for running a specific task. See the `r:` group definition near the beginning of the [crossbow configuration](https://github.com/apache/arrow/blob/master/dev/tasks/tasks.yml) for a list of glob expression patterns that match names of items in the `tasks:` list below it.
```

nealrichardson commented on r/vignettes/developing.Rmd:
```diff
@@ -0,0 +1,510 @@
+
+## Github workflows
+
+Crossbow is our extended test suite runner [more documentation is available](https://arrow.apache.org/docs/developers/crossbow.html) but important GitHub comment commands include:
```
Restructure this more like "On a pull request, there are some actions you can trigger by commenting on the PR. We have additional CI checks that run nightly and can be requested on demand using an internal tool called [crossbow](link)."

nealrichardson commented on r/vignettes/install.Rmd:
```diff
@@ -335,6 +286,16 @@ See discussion [here](https://issues.apache.org/jira/browse/ARROW-8556).
 
+Some features are optional when you build Arrow from source
+
+* `ARROW_S3`: If set to `true` S3 support will be built (as long as the 
+  dependencies are met, if they are not met, this may be automatically turned off) (default: false)
+* `ARROW_JEMALLOC`: If set to `true` jemalloc support will be included (default: true)
+* `ARROW_PARQUET`: If set to `true` parquet support will be included (default: true)
+* `ARROW_DATASET`: If set to `true` dataset support will be included (default: true)
+* `ARROW_WITH_RE2`: If set to `true` RE2 regular expression library support will be included (default: true)
+* `ARROW_WITH_UTF8PROC`: If set to `true` UTF8Proc string library support will be included (default: true)
```
```suggestion
* `ARROW_S3`: If set to `ON` S3 support will be built as long as the 
  dependencies are met; if they are not met, the build script will turn this `OFF` 
* `ARROW_JEMALLOC` for the `jemalloc` memory allocator
* `ARROW_PARQUET`
* `ARROW_DATASET`
* `ARROW_WITH_RE2` for the RE2 regular expression library, used in some string compute functions
* `ARROW_WITH_UTF8PROC` for the UTF8Proc string library, used in many other string compute functions
```

nealrichardson commented on r/vignettes/install.Rmd:
```diff
@@ -335,6 +286,16 @@ See discussion [here](https://issues.apache.org/jira/browse/ARROW-8556).
 
 ## Summary of build environment variables
 
+Some features are optional when you build Arrow from source
```
```suggestion
Some features are optional when you build Arrow from source. With the exception of `ARROW_S3`, these are all `ON` by default in the bundled C++ build, but you can set them to `OFF` to disable them.
```

nealrichardson commented on r/vignettes/install.Rmd:
```diff
@@ -335,6 +286,16 @@ See discussion [here](https://issues.apache.org/jira/browse/ARROW-8556).
+* `ARROW_WITH_RE2`: If set to `true` RE2 regular expression library support will be included (default: true)
+* `ARROW_WITH_UTF8PROC`: If set to `true` UTF8Proc string library support will be included (default: true)
+
 By default, these are all unset. All boolean variables are case-insensitive.
```
```suggestion
There are a number of other variables that affect the `configure` script and the bundled build script.
By default, these are all unset. All boolean variables are case-insensitive.
```

nealrichardson commented on r/vignettes/developing.Rmd:
````diff
@@ -0,0 +1,520 @@
+brew install apache-arrow --HEAD
+```
+
+On Windows and Linux, you can download a .zip file with the arrow dependencies from the
+nightly repository,
+and then set the `RWINLIB_LOCAL` environment variable to point to that
+zip file before installing the `arrow` R package. Version numbers in that
````
```suggestion
On Windows and Linux, you can download a .zip file with the arrow dependencies from the
nightly repository.
Windows users then can set the `RWINLIB_LOCAL` environment variable to point to that
zip file before installing the `arrow` R package. On Linux, you'll need to create a `libarrow` directory inside the R package directory and unzip that file into it. Version numbers in that
```

nealrichardson replied:
(This is lame and we could do a lot better, but at least we should be factually accurate.)
